File: implementacion_java/src/algoritmos/WinogradScaled.java
package algoritmos;

import utilidades.AlgoritmoMultiplicacion;
import utilidades.Convertidor;

import java.math.BigInteger;

public class WinogradScaled implements AlgoritmoMultiplicacion {

    Convertidor convertidor = new Convertidor();


    public static double NormInf(double[][] A, int N, int M) {
        // Inicializa la variable max con el valor más bajo posible
        double max = Double.NEGATIVE_INFINITY;

        // Recorre cada fila de la matriz
        for (int i = 0; i < N; i++) {
            double sum = 0.0; // Inicializa la suma de elementos absolutos de la fila actual

            // Recorre cada columna de la fila actual
            for (int j = 0; j < M; j++) {
                sum += Math.abs(A[i][j]); // Suma el valor absoluto del elemento actual
            }

            // Actualiza el valor máximo si la suma actual es mayor que el máximo registrado
            if (sum > max) {
                max = sum;
            }
        }

        // Devuelve la suma máxima de elementos absolutos por fila (norma infinito)
        return max;
    }

    public static void MultiplyWithScalar(double[][] A, double[][] B, int N, int M, double scalar) {
        // Recorre cada elemento de la matriz
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                B[i][j] = A[i][j] * scalar; // Multiplica el elemento por el escalar y lo almacena en la matriz de salida
            }
        }
    }

    @Override
    public void multiplicar(BigInteger[][] matriz1, BigInteger[][] matriz2) {
        // Convertir de BigInteger a double
        double[][] doubleMatriz1 = convertidor.convertBigIntegerArrayToDoubleArray(matriz1);
        double[][] doubleMatriz2 = convertidor.convertBigIntegerArrayToDoubleArray(matriz2);
        double[][] doubleResultado = multiplicarWinogradScaled(doubleMatriz1, doubleMatriz2);

        // Convertir el resultado de double a BigInteger
        convertidor.convertDoubleArrayToBigInteger(doubleResultado);
    }

    public double[][] multiplicarWinogradScaled(double[][] A, double[][] B) {
        int N = A.length; // Número de filas de A
        int P = A[0].length; // Número de columnas de A (y filas de B)
        int M = B[0].length; // Número de columnas de B
        double[][] C = new double[N][M]; // Matriz de resultado

        // Crear copias escaladas de A y B
        double[][] CopyA = new double[N][P];
        double[][] CopyB = new double[P][M];

        // Factores de escalado
        double a = NormInf(A, N, P); // Norma infinito de A
        double b = NormInf(B, P, M); // Norma infinito de B
        double lambda = Math.floor(0.5 + Math.log(b / a) / Math.log(4)); // Cálculo del factor de escalado

        // Escalar las matrices A y B
        MultiplyWithScalar(A, CopyA, N, P, Math.pow(2, lambda));
        MultiplyWithScalar(B, CopyB, P, M, Math.pow(2, -lambda));

        // Usar Winograd con matrices escaladas
        C = multiplyWinograd(CopyA, CopyB);

        return C; // Devolver la matriz resultado
    }

    /**
     * Multiplica matrices usando el algoritmo de Winograd.
     */
    private double[][] multiplyWinograd(double[][] A, double[][] B) {
        int N = A.length; // Número de filas de A
        int M = B[0].length; // Número de columnas de B
        int P = B.length; // Número de columnas de A (y filas de B)
        double[][] C = new double[N][M]; // Matriz de resultado

        // Variables intermedias para el algoritmo de Winograd
        double[] rowFactor = new double[N];
        double[] colFactor = new double[M];

        // Precalcular factores de filas
        for (int i = 0; i < N; i++) {
            rowFactor[i] = A[i][0] * A[i][1]; // Multiplicar los dos primeros elementos de cada fila
            for (int j = 2; j < P; j += 2) {
                rowFactor[i] += A[i][j] * A[i][j + 1]; // Sumar productos de pares de elementos
            }
        }

        // Precalcular factores de columnas
        for (int j = 0; j < M; j++) {
            colFactor[j] = B[0][j] * B[1][j]; // Multiplicar los dos primeros elementos de cada columna
            for (int i = 2; i < P; i += 2) {
                colFactor[j] += B[i][j] * B[i + 1][j]; // Sumar productos de pares de elementos
            }
        }

        // Realizar la multiplicación de matrices
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                C[i][j] = -rowFactor[i] - colFactor[j]; // Inicializar con la suma de factores negativos
                for (int k = 0; k < P; k += 2) {
                    C[i][j] += (A[i][k] + B[k + 1][j]) * (A[i][k + 1] + B[k][j]); // Sumar productos ajustados
                }
            }
        }

        // Ajustar si la dimensión es impar
        if (P % 2 == 1) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    C[i][j] += A[i][P - 1] * B[P - 1][j]; // Añadir el último producto si P es impar
                }
            }
        }

        return C; // Devolver la matriz resultado
    }
}

File: implementacion_kotlin/src/multiplicacion_matrices/WinogradScaled.kt
package multiplicacion_matrices

import utilidades.AlgoritmoMultiplicacion
import utilidades.Convertidor
import java.math.BigInteger

class WinogradScaled : AlgoritmoMultiplicacion {

    private val convertidor = Convertidor()

    override fun multiplicar(matriz1: Array<Array<BigInteger>>, matriz2: Array<Array<BigInteger>>) {
        // Convertir de BigInteger a double
        val doubleMatriz1 = convertidor.convertBigIntegerArrayToDoubleArray(matriz1)
        val doubleMatriz2 = convertidor.convertBigIntegerArrayToDoubleArray(matriz2)
        val doubleResultado = multiplicarWinogradScaled(doubleMatriz1, doubleMatriz2)

        // Convertir el resultado double a BigINteger
        convertidor.convertDoubleArrayToBigInteger(doubleResultado)
    }

    private fun multiplicarWinogradScaled(A: Array<DoubleArray>, B: Array<DoubleArray>): Array<DoubleArray> {
        val N = A.size
        val P = A[0].size
        val M = B[0].size
        val C: Array<DoubleArray>

        // Create scaled copies of A and B
        val copyA = Array(N) { DoubleArray(P) }
        val copyB = Array(P) { DoubleArray(M) }

        // Scaling factors
        val a = NormInf(A, N, P)
        val b = NormInf(B, P, M)
        val lambda = Math.floor(0.5 + Math.log(b / a) / Math.log(4.0))

        // Scaling
        MultiplyWithScalar(A, copyA, N, P, Math.pow(2.0, lambda))
        MultiplyWithScalar(B, copyB, P, M, Math.pow(2.0, -lambda))

        // Using Winograd with scaled matrices
        C = multiplyWinograd(copyA, copyB)

        return C
    }

    /**
     * Multiplies matrices using Winograd algorithm.
     */
    private fun multiplyWinograd(A: Array<DoubleArray>, B: Array<DoubleArray>): Array<DoubleArray> {
        val N = A.size
        val M = B[0].size
        val P = B.size
        val C = Array(N) { DoubleArray(M) }

        // Intermediate variables for Winograd algorithm
        val rowFactor = DoubleArray(N)
        val colFactor = DoubleArray(M)

        // Precompute row factors
        for (i in 0 until N) {
            rowFactor[i] = A[i][0] * A[i][1]
            for (j in 2 until P step 2) {
                rowFactor[i] += A[i][j] * A[i][j + 1]
            }
        }

        // Precompute column factors
        for (j in 0 until M) {
            colFactor[j] = B[0][j] * B[1][j]
            for (i in 2 until P step 2) {
                colFactor[j] += B[i][j] * B[i + 1][j]
            }
        }

        // Actual matrix multiplication
        for (i in 0 until N) {
            for (j in 0 until M) {
                C[i][j] = -rowFactor[i] - colFactor[j]
                for (k in 0 until P step 2) {
                    C[i][j] += (A[i][k] + B[k + 1][j]) * (A[i][k + 1] + B[k][j])
                }
            }
        }

        // Adjust for odd dimension
        if (P % 2 == 1) {
            for (i in 0 until N) {
                for (j in 0 until M) {
                    C[i][j] += A[i][P - 1] * B[P - 1][j]
                }
            }
        }

        return C
    }

    /**
     * Calcula la norma infinito de una matriz.
     *
     *@param A Matriz de entrada.
     * @param N Número de filas.
     * @param M Número de columnas.
     * @return La norma infinito de la matriz.
     */
    fun NormInf(A: Array<DoubleArray>, N: Int, M: Int): Double {
        var max = Double.NEGATIVE_INFINITY
        for (i in 0 until N) {
            var sum = 0.0
            for (j in 0 until M) {
                sum += Math.abs(A[i][j])
            }
            if (sum > max) {
                max = sum
            }
        }
        return max
    }

    /**
     * Multiplica una matriz por un escalar.
     *
     * @param A      Matriz de entrada.
     * @param B      Matriz de salida.
     * @param N      Número de filas.
     * @param M      Número de columnas.
     * @param scalar Escalar.
     */
    fun MultiplyWithScalar(A: Array<DoubleArray>, B: Array<DoubleArray>, N: Int, M: Int, scalar: Double) {
        for (i in 0 until N) {
            for (j in 0 until M) {
                B[i][j] = A[i][j] * scalar
            }
        }
    }


}